core/src/main/scala-3/chisel3/VerificationStatementIntf.scala

```diff
@@ -2,14 +2,65 @@
 
 package chisel3
 
-private[chisel3] trait Assert$Intf { self: assert.type =>
-  // TODO implement
+import chisel3.experimental.{BaseModule, SourceInfo}
+
+trait VerifPrintMacrosDoc
+
+val emptySourceLine: VerifStmtMacrosCompat.SourceLineInfo = ("???", 0)
+
+private[chisel3] trait Assert$Intf extends VerifPrintMacrosDoc { self: assert.type =>
+
+  def apply(
+    cond:    Bool,
+    message: String,
+    data:    Bits*
+  )(
+    implicit sourceInfo: SourceInfo
+  ): Assert = VerifStmtMacrosCompat.assert._applyWithSourceLinePrintable(
+    cond,
+    emptySourceLine,
+    Some(Printable.pack(message, data: _*))
+  )
+
+  def apply(
+    cond:    Bool,
+    message: Printable
+  )(
+    implicit sourceInfo: SourceInfo
+  ): Assert = VerifStmtMacrosCompat.assert._applyWithSourceLinePrintable(cond, emptySourceLine, Some(message))
+
+  def apply(cond: Bool)(implicit sourceInfo: SourceInfo): Assert =
+    VerifStmtMacrosCompat.assert._applyWithSourceLinePrintable(cond, emptySourceLine, None)
 }
```

Comment:
Does this need to be two methods, one that is variadic and one that isn't? Can this just be one method with branching in the body based on `data.nonEmpty`?

Reply:
Same question for later methods.

Reply:
Yeah there's a bunch of cleanup/consolidation possible here that I was tempted to do (maybe even with Scala 3's Union types?). Decided to keep this as-is for now so that it's laid out the same as in Scala 2 which should make it a bit easier to reason about when all this eventually gets replaced with the macro versions, which I think is a better point to work on cleaning it up